internal/bufferedwrites/buffered_write_handler_test.go

```diff
@@ -186,5 +187,61 @@ func (testSuite *BufferedWriteTest) TestFlush_SignalUploadFailureDuringWrite() {
 	obj, err := testSuite.bwh.Flush()
 	require.Error(testSuite.T(), err)
 	assert.Equal(testSuite.T(), err, ErrUploadFailure)
-	assert.Nil(testSuite.T(), obj)
+	// Whatever could be finalized, got finalized (empty object in this case).
+	assert.NotNil(testSuite.T(), obj)
+	assert.Equal(testSuite.T(), uint64(0), obj.Size)
+}
+
+func (testSuite *BufferedWriteTest) TestFlushWithMultiBlockWritesAndSignalUploadFailureInBetween() {
+	buffer, err := operations.GenerateRandomData(blockSize)
+	assert.NoError(testSuite.T(), err)
+	// Upload and sync 5 blocks.
+	testSuite.TestSync5InProgressBlocks()
+	// Close the channel to simulate failure in uploader.
+	close(testSuite.bwh.uploadHandler.SignalUploadFailure())
+	// Write 5 more blocks.
+	for i := 0; i < 5; i++ {
+		err := testSuite.bwh.Write(buffer, int64(blockSize*(i+5)))
+		require.Error(testSuite.T(), err)
+	}
+
+	obj, err := testSuite.bwh.Flush()
+
+	require.Error(testSuite.T(), err)
+	assert.Equal(testSuite.T(), err, ErrUploadFailure)
+	// Whatever could be finalized, got finalized.
+	assert.NotNil(testSuite.T(), obj)
+	assert.Equal(testSuite.T(), uint64(5*blockSize), obj.Size)
+}
+
+func (testSuite *BufferedWriteTest) TestSync5InProgressBlocks() {
+	buffer, err := operations.GenerateRandomData(blockSize)
+	assert.NoError(testSuite.T(), err)
+	// Write 5 blocks.
+	for i := 0; i < 5; i++ {
+		err = testSuite.bwh.Write(buffer, int64(blockSize*i))
+		require.Nil(testSuite.T(), err)
+	}
+
+	// Wait for 5 blocks to upload successfully.
+	err = testSuite.bwh.Sync()
+
+	assert.NoError(testSuite.T(), err)
```

Comment:
is it possible to check if all blocks are uploaded from the uploadHandler.uploadCh?

Reply:
Added these two checks: 
```
	assert.Equal(testSuite.T(), 0, len(testSuite.bwh.uploadHandler.uploadCh))
	assert.Equal(testSuite.T(), 5, len(testSuite.bwh.blockPool.FreeBlocksChannel()))
```